Neos.Flow/Classes/Http/BaseRequest.php

```diff
@@ -0,0 +1,389 @@
+<?php
+namespace Neos\Flow\Http;
+
+use Psr\Http\Message\RequestInterface;
+use Psr\Http\Message\UriInterface;
+use Neos\Utility\MediaTypes;
+
+/**
+ *
+ */
+class BaseRequest extends AbstractMessage implements RequestInterface
+{
+    /**
+     * @var Uri
+     */
+    protected $uri;
+
+    /**
+     * @var Uri
+     */
+    protected $baseUri;
+
+    /**
+     * @var string
+     */
+    protected $method = 'GET';
+
+    /**
+     * URI for the "input" stream wrapper which can be modified for testing purposes
+     *
+     * @var string
+     */
+    protected $inputStreamUri = null;
+
+    /**
+     * @var string
+     */
+    protected $requestTarget = null;
+
+    /**
+     * BaseRequest constructor.
+     *
+     * @param UriInterface $uri
+     * @param string $method
+     */
+    public function __construct(UriInterface $uri, $method = 'GET')
+    {
+        $this->uri = $uri;
+        $this->method = $method;
+    }
+
+    /**
+     * Returns the request URI
+     *
+     * @return Uri
+     * @api
+     */
+    public function getUri()
+    {
+        return $this->uri;
+    }
+
+    /**
+     * Returns the detected base URI
+     *
+     * @return Uri
+     * @api
+     */
+    public function getBaseUri()
+    {
+        if ($this->baseUri === null) {
+            $this->detectBaseUri();
+        }
+
+        return $this->baseUri;
+    }
+
+    /**
+     * @param Uri $baseUri
+     */
+    public function setBaseUri($baseUri)
+    {
+        $this->baseUri = $baseUri;
+    }
+
+    /**
+     * Tries to detect the base URI of request.
+     *
+     * @return void
```

Comment:
I think the `if` clause should be outside this method, because running `detectBaseUri` and then not detecting the base uri is somewhat unexpected. Even better would be if this method would return the uri instead of setting it.